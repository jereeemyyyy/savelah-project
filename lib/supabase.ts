import { AppState } from 'react-native'
import 'react-native-url-polyfill/auto'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = 'https://cjcldaiiqmedjmojjzfz.supabase.co'
const supabaseAnonKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImNqY2xkYWlpcW1lZGptb2pqemZ6Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MTY3OTE2NDksImV4cCI6MjAzMjM2NzY0OX0.6z2M2mG_qmSPPwujMWYpN4KkLgiU0mWjGdsCdVZvlNw'

export const supabase = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    storage: AsyncStorage,
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: false,
  },
})

// Tells Supabase Auth to continuously refresh the session automatically
// if the app is in the foreground. When this is added, you will continue
// to receive `onAuthStateChange` events with the `TOKEN_REFRESHED` or
// `SIGNED_OUT` event if the user's session is terminated. This should
// only be registered once.
AppState.addEventListener('change', (state) => {
  if (state === 'active') {
    supabase.auth.startAutoRefresh()
  } else {
    supabase.auth.stopAutoRefresh()
  }
})