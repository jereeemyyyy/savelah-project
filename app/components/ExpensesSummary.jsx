import React, { useState, useEffect } from 'react';
import { View, Text, Dimensions, SafeAreaView, TouchableOpacity } from 'react-native';
import { PieChart } from 'react-native-chart-kit';
import { supabase } from '../../lib/supabase';
import getColour from '../utils/getColour';

const screenWidth = Dimensions.get('window').width;

export default function ExpensesSummary({ userId }) {
  const [data, setData] = useState([]);
  const [currentWeekTotal, setCurrentWeekTotal] = useState(0);
  const [lastWeekTotal, setLastWeekTotal] = useState(0);

  const fetchExpensesByCategory = async () => {
    try {
      const { data: fetchedData, error } = await supabase.rpc('get_total_expenses_by_category', {
        p_user_id: userId
      });

      if (error) {
        console.log(error);
        throw new Error('Error fetching expenses by category');
      }

      // Map the fetched data to the format expected by PieChart and validate it
      const formattedData = fetchedData.map(item => {
        const amount = parseFloat(item.total_expense);
        const color = getColour(item.category);

        return {
          name: item.category,
          population: item.total_expense,
          color,
          legendFontColor: '#000000',
          legendFontSize: 12
        };
      }).filter(item => item !== null); // Remove invalid entries

      setData(formattedData);
    } catch (error) {
      console.error('Error fetching expenses by category:', error.message);
    }
  };

  const fetchWeeklyTotals = async () => {
    try {
      const { data: currentWeekData, error: currentWeekError } = await supabase.rpc('get_total_expenses_current_week', {
        p_user_id: userId
      });

      if (currentWeekError) {
        console.error('Error fetching current week expenses:', currentWeekError);
        return;
      }

      setCurrentWeekTotal(currentWeekData);

      const { data: lastWeekData, error: lastWeekError } = await supabase.rpc('get_total_expenses_last_week', {
        p_user_id: userId
      });

      if (lastWeekError) {
        console.error('Error fetching last week expenses:', lastWeekError);
        return;
      }

      setLastWeekTotal(lastWeekData);
    } catch (error) {
      console.error('Error fetching weekly totals:', error.message);
    }
  };

  useEffect(() => {
    if (userId) {
      fetchExpensesByCategory();
      fetchWeeklyTotals();
    }

    const subscription = supabase
      .channel('room1')
      .on('postgres_changes', { event: '*', schema: 'public', table: 'categories', filter: `user_id=eq.${userId}` }, payload => {
        fetchExpensesByCategory();
        fetchWeeklyTotals();
      })
      .subscribe();

    return () => {
      supabase.removeChannel(subscription);
    };
  }, [userId]);

  const handleRefresh = () => {
    fetchExpensesByCategory();
    fetchWeeklyTotals();
  };

  return (
    <SafeAreaView className="flex-1 justify-start items-center">
      <TouchableOpacity className="bg-white rounded-lg p-5 shadow-lg w-11/12" onPress={handleRefresh}>
        {data.length > 0 ? (
          <PieChart
            data={data.map(item => ({
              name: item.name,
              population: item.population,
              color: item.color,
              legendFontColor: item.legendFontColor,
              legendFontSize: item.legendFontSize
            }))}
            width={screenWidth - 80}
            height={150}
            chartConfig={{
              backgroundColor: '#ffffff',
              backgroundGradientFrom: '#ffffff',
              backgroundGradientTo: '#ffffff',
              color: (opacity = 1) => `rgba(0, 0, 0, ${opacity})`,
            }}
            accessor="population"
            backgroundColor="transparent"
            paddingLeft="10"
            paddingRight="0"
            absolute
          />
        ) : (
          <Text className="text-gray-600 mb-4">No expenses data available.</Text>
        )}

        <View className="mt-1">
          <Text className="text-base text-black font-bold">Last Week: ${lastWeekTotal}</Text>
          <Text className="text-base text-black font-bold">This Week: ${currentWeekTotal}</Text>
        </View>
      </TouchableOpacity>
    </SafeAreaView>
  );
}
