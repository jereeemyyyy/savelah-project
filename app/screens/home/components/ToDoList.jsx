import React, { useState, useEffect } from 'react';
import { View, Text, TouchableOpacity, FlatList, SafeAreaView } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import AddTaskButton from './AddTaskButton';
import { supabase } from '../../../../lib/supabase';
import ConfirmationModal from './ConfirmationModal'; // Import the modal component
import * as eva from '@eva-design/eva';
import { ApplicationProvider } from '@ui-kitten/components';

const defaultCategories = ['Food', 'Transport', 'Housing'];

export default function ToDoList() {
  const [tasks, setTasks] = useState([]);
  const [loading, setLoading] = useState(true);
  const [modalVisible, setModalVisible] = useState(false);
  const [selectedTask, setSelectedTask] = useState(null);

  const fetchTasks = async () => {
    setLoading(true);
    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (user) {
        const { data, error } = await supabase
          .from('to_do_list')
          .select('*')
          .eq('user_id', user.id);
        if (error) {
          console.error('Error fetching tasks:', error);
        } else {
          const formattedTasks = data.map((task) => ({
            id: task.id,
            title: task.title,
            description: task.description,
            amount: task.amount_spent,
            time: task.time 
          }));
          setTasks(formattedTasks);
        }
      }
    } catch (error) {
      console.error('Unexpected error fetching tasks:', error);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchTasks();
  }, []);

  const addTask = (task) => {
    setTasks([...tasks, task]);
  };

  const handleAddToConfirmation = (task) => {
    setSelectedTask(task);
    setModalVisible(true);
  };

  const renderItem = ({ item }) => (
    <View className="bg-white p-4 rounded-lg shadow-md mb-4 mx-4">
      <View className="flex-row items-center justify-between">
        <View className="flex-3">
          <Text className="text-lg font-bold">{item.title}</Text>
          <Text>{item.description}</Text>
          <Text className="text-sm text-gray-500">{new Date(item.time).toLocaleString()}</Text>
        </View>
        <View className="flex-1 items-center">
          <Text className={`font-bold text-lg ${item.amount >= 0 ? 'text-green-500' : 'text-red-500'}`}>
            ${item.amount}
          </Text>
        </View>
        <TouchableOpacity
          onPress={() => handleAddToConfirmation(item)}
          className="bg-purple-500 p-3 rounded"
        >
          <Text className="text-white font-bold">Add</Text>
        </TouchableOpacity>
      </View>
    </View>
  );

  return (
    <SafeAreaView className="flex-1">
      <View className="flex-row bg-gray-800 p-4 justify-between items-center">
        <Text className="text-3xl font-bold text-white mb-1">To-Do List</Text>
          <AddTaskButton 
            addTask={addTask} 
            fetchTasks={fetchTasks}
          />
          <TouchableOpacity
            onPress={fetchTasks}
            className="bg-purple-500 p-3 rounded -ml-8"
          >
            <Icon name="refresh" size={22} color="white" />
          </TouchableOpacity>
        
      </View>
      {loading ? (
        <View className="flex-1 justify-center items-center">
          <Text>Loading...</Text>
        </View>
      ) : (
        <FlatList
          data={tasks}
          renderItem={renderItem}
          keyExtractor={(item) => item.id.toString()}
          contentContainerClassName="p-4"
        />
      )}
      {selectedTask && (
        <ApplicationProvider {...eva} theme={eva.light}>
          <ConfirmationModal
            task={selectedTask}
            modalVisible={modalVisible}
            setModalVisible={setModalVisible}
            fetchTasks={fetchTasks}
          />
        </ApplicationProvider>
      )}
    </SafeAreaView>
  );
}
