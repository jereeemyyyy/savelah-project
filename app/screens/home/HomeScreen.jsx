import { Text, View } from 'react-native';
import  BackButton  from '../../components/BackButton';
import  ExpensesSummary from './components/ExpensesSummary';

export default function HomeScreen() {
    return (
        <View className="justify-start bg-gray-800 flex-1 flex-col">
            <Text className="font-bold text-xl p-5 text-white">Home Screen</Text>   
            <View className="flex-1">
                <ExpensesSummary/>
            </View>  
            <Text className="font-bold text-xl p-5 text-white">To Do List</Text>   
        </View>
    );  
    
}