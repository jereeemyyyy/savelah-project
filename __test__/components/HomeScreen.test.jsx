import React from 'react';
import { Text, View, SafeAreaView } from 'react-native';
import renderer, { act } from 'react-test-renderer';
import HomeScreen from '../../app/screens/home/HomeScreen';
import { supabase } from '../../lib/supabase';
import ExpensesSummary from '../../app/components/ExpensesSummary';
import ToDoList from '../../app/screens/home/components/ToDoList';

jest.mock('../../lib/supabase', () => ({
    supabase: {
        auth: {
            getUser: jest.fn(),
        },
    },
}));

jest.mock('../../app/components/ExpensesSummary', () => 'ExpensesSummary');
jest.mock('../../app/screens/home/components/ToDoList', () => 'ToDoList');

describe('HomeScreen', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let tree;
        await act(async () => {
            tree = renderer.create(<HomeScreen />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('contains the expected text elements', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let component;
        await act(async () => {
            component = renderer.create(<HomeScreen />);
        });
        const instance = component.root;

        expect(instance.findByType(Text).props.children).toContain('Expenses This Week');
    });

    it('contains the ExpensesSummary and ToDoList components', async () => {
        supabase.auth.getUser.mockResolvedValue({ data: { user: { id: '123' } } });

        let component;
        await act(async () => {
            component = renderer.create(<HomeScreen />);
        });
        const instance = component.root;

        expect(instance.findByType('ExpensesSummary')).toBeTruthy();
        expect(instance.findByType('ToDoList')).toBeTruthy();
    });
});
