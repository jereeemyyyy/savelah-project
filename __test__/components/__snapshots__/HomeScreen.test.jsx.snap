// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#1f2937",
      },
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "fontWeight": "700",
          },
          {
            "fontSize": 30,
            "lineHeight": 36,
          },
          {
            "color": "#fff",
          },
        ]
      }
    >
      Expenses This Week
    </Text>
  </View>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "marginBottom": -40,
        },
      ]
    }
  >
    <ExpensesSummary
      userId="123"
    />
  </View>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "marginTop": -80,
        },
      ]
    }
  >
    <ToDoList />
  </View>
</RCTSafeAreaView>
`;
